### PR TITLE
Add post handshake authentication extension 

### PR DESCRIPTION
**Pull request by smittal3**
### Description of changes: 
Added pha_ext extension to SSL and SSL_CTX for post handshake authentication
Added functions for setting pha_ext in SSL and SSL_CTX
Added logic for pha_ext to add/parse to/from Client Hello
Added stubs for all relevant functions related to PHA
Added SSL_VERIFY_POST_HANDSHAKE macro to delay client authentication to after the initial handshake and corresponding logic

### Testing:
So far unit tests exist for setting pha_ext in SSL and SSL_CTX, ensuring it is sent by client and received by server, and verifying behavior when SSL_VERIFY_POST_HANDSHAKE is set 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and the ISC license.

### CODE REVIEW

File: include/openssl/ssl.h [hunk 1]
```diff
@@ -4501,6 +4506,30 @@ OPENSSL_EXPORT int SSL_CTX_set_max_send_fragment(SSL_CTX *ctx,
...
                                              size_t max_send_fragment);
 
+
+
+// SSL_CTX_set_post_handshake_auth and SSL_set_post_handshake_auth enable the
+// Post-Handshake Authentication extension (pha_ext) to be added to the ClientHello
+// such that post-handshake authentication can be requested by the server.
+// If val is 0 then the extension is not sent, if it is 1 then it is.
+// By default, the extension is not sent. On the client side, a certificate callback
+// will need to be set via SSL_CTX_set_client_cert_cb() if no certificate is
```
Comment (samuel40791765) [inline include/openssl/ssl.h]: Nit: Just our typical coding style
```suggestion
// will need to be set via |SSL_CTX_set_client_cert_cb| if no certificate is
```

File: include/openssl/ssl.h [hunk 2]
```diff
@@ -4501,6 +4506,30 @@ OPENSSL_EXPORT int SSL_CTX_set_max_send_fragment(SSL_CTX *ctx,
...
+// provided at initialization.
+OPENSSL_EXPORT void SSL_CTX_set_post_handshake_auth(SSL_CTX *ctx, int val);
+OPENSSL_EXPORT void SSL_set_post_handshake_auth(SSL *ssl, int val);
+
+// Causes a CertificateRequest message to be sent by a server on the given ssl connection.
+// The SSL_VERIFY_PEER flag must be set; the SSL_VERIFY_POST_HANDSHAKE flag is optional.
+OPENSSL_EXPORT int SSL_verify_client_post_handshake(SSL *ssl);
```
Comment (samuel40791765) [inline include/openssl/ssl.h]:
```suggestion
// SSL_verify_client_post_handshake lets the server send a CertificateRequest message on the given ssl connection.
// The |SSL_VERIFY_PEER| flag must be set; the |SSL_VERIFY_POST_HANDSHAKE| flag is optional.
OPENSSL_EXPORT int SSL_verify_client_post_handshake(SSL *ssl);
```

Comment (samuel40791765) [reply include/openssl/ssl.h]: I can see how it's done in the tests, but it would be nice to know how to set these two flags in the documentation.

File: include/openssl/ssl.h [hunk 3]
```diff
@@ -4501,6 +4506,30 @@ OPENSSL_EXPORT int SSL_CTX_set_max_send_fragment(SSL_CTX *ctx,
...
+
+
+// Returns the status of post handshake authentication. Returns X509_V_OK if
+// peer certificate is provided and valid and NULL otherwise. The function calls
+// SSL_get_peer_certificate to ensure a peer certificate has been provided
+// and SSL_get_verify_result to ensure the provided certificate is valid
```
Comment (samuel40791765) [inline include/openssl/ssl.h]:
```suggestion
// SSL_get_verify_result_pha returns the status of post handshake authentication. It 
// returns |X509_V_OK| if the peer certificate is provided and valid, and NULL if otherwise.
// The function calls |SSL_get_peer_certificate| to ensure a peer certificate has been provided
// and |SSL_get_verify_result| to ensure the provided certificate is valid.
```

Comment (samuel40791765) [reply include/openssl/ssl.h]: I'm not quite sure if this documentation is valid, `SSL_get_verify_result_pha` returns a `true` and `false` instead of what's being said here

File: ssl/ssl_x509.cc [hunk 4]
```diff
@@ -560,6 +560,15 @@ BSSL_NAMESPACE_END
 
 using namespace bssl;
 
+// Returns the status of PHA based on results of |SSL_get_peer_certificate| and
+// |SSL_get_verify_result|
```
Comment (samuel40791765) [inline ssl/ssl_x509.cc]: Comment is a bit redundant, since you have nice documentation in ssl.h already

File: ssl/ssl_x509.cc [hunk 5]
```diff
@@ -560,6 +560,15 @@ BSSL_NAMESPACE_END
...
+long SSL_get_verify_result_pha(const SSL *ssl) {
+  if(SSL_get_peer_certificate(ssl) != NULL && SSL_get_verify_result(ssl) == X509_V_OK) {
```
Comment (samuel40791765) [inline ssl/ssl_x509.cc]: Minor nit in C++:
```suggestion
  if(SSL_get_peer_certificate(ssl) != nullptr && SSL_get_verify_result(ssl) == X509_V_OK) {
```

File: include/openssl/ssl.h [hunk 6]
```diff
@@ -4610,6 +4639,17 @@ OPENSSL_EXPORT int SSL_was_key_usage_invalid(const SSL *ssl);
...
+
+// SSL_PHA_STATE enumerates possible Post Handshake Authentication states
+typedef enum {
+  SSL_PHA_NONE = 0,
+  SSL_PHA_ENABLED = 1,
+  SSL_PHA_EXT_SENT = 2,        /* client-side only: extension sent */
+  SSL_PHA_EXT_RECEIVED = 3,    /* server-side only: extension received */
+  SSL_PHA_REQUESTED = 4,       /* server-side only: request sent by server */
+  SSL_PHA_REQUEST_PENDING = 5 /* server-side only: request needs to be sent by server */
+} SSL_PHA_STATE;
```
Comment (samuel40791765) [inline include/openssl/ssl.h]: OpenSSL seems to place this in a local header file, which is our `internal.h`. 
It's important for us not to expose any unnecessary symbols externally :)

File: ssl/ssl_lib.cc [hunk 7]
```diff
@@ -929,6 +931,24 @@ static int ssl_do_post_handshake(SSL *ssl, const SSLMessage &msg) {
...
 }
 
+// Sets pha_ext value in SSL_CTX
+void SSL_CTX_set_post_handshake_auth(SSL_CTX *ctx, int val) {
+  ctx->pha_ext = val;
+}
+
+// Sets pha_ext value in SSL
+void SSL_set_post_handshake_auth(SSL *ssl, int val) {
+  ssl->s3->pha_ext = val;
```
Comment (samuel40791765) [inline ssl/ssl_lib.cc]: Won't setting these flip the internal state of the post handshake authentication? 
This doesn't seem to be part of the function contract and the internal state of PHA shouldn't be exposed for the user to configure. We should also separate the configuration and the state management of PHA into different fields. OpenSSL also does this with two variables, one for enabling pha, one for maintaining the state.
https://github.com/openssl/openssl/blob/9c8d04dbec03172d6ffe4eaa38ea4b1ac2741f26/ssl/ssl_local.h#L1725-L1726C9

You can consider using a `bool` internally to maintain whether PHA is enabled or not. Something clear like `post_handshake_state` and `post_handshake_enabled` would be nice :)

File: ssl/extensions.cc [hunk 8]
```diff
@@ -3089,6 +3089,48 @@ bool ssl_negotiate_alps(SSL_HANDSHAKE *hs, uint8_t *out_alert,
...
+  if(hs->min_version < TLS1_3_VERSION || ssl->s3->pha_ext != SSL_PHA_ENABLED) {
+    return true;
+  }
+
+  CBB contents;
+  if (!CBB_add_u16(out_compressible, TLSEXT_TYPE_post_handshake_auth) ||
+      !CBB_add_u16_length_prefixed(out_compressible, &contents) ||
+      !CBB_add_u16(&contents, 0 /* empty "extension_data" field */) ||
+      !CBB_flush(out_compressible)) {
+    return false;
```
Comment (samuel40791765) [inline ssl/extensions.cc:3106]: Should we throw an error on the stack here? I noticed OpenSSL does this.
https://github.com/openssl/openssl/blob/master/ssl/statem/extensions_clnt.c#L1203

File: ssl/extensions.cc [hunk 9]
```diff
@@ -3089,6 +3089,48 @@ bool ssl_negotiate_alps(SSL_HANDSHAKE *hs, uint8_t *out_alert,
...
+
+  ssl->s3->pha_ext = SSL_PHA_EXT_SENT;
+  return true;
+}
+
+// Parses pha_ext from client hello
+static bool ext_pha_parse_clienthello(SSL_HANDSHAKE *hs,
+                                      uint8_t *out_alert,
+                                      CBS *contents)  {
+  if(contents == NULL || hs->min_version < TLS1_3_VERSION) {
```
Comment (samuel40791765) [inline ssl/extensions.cc]: Nit:
```suggestion
  if(contents == nullptr || hs->min_version < TLS1_3_VERSION) {
```

File: ssl/extensions.cc [hunk 10]
```diff
@@ -3089,6 +3089,48 @@ bool ssl_negotiate_alps(SSL_HANDSHAKE *hs, uint8_t *out_alert,
...
+    return true;
+  }
+
+  if(CBS_len(contents) == 0) {
+    return false;
+  }
+
+  const SSL *const ssl = hs->ssl;
+  ssl->s3->pha_ext = SSL_PHA_EXT_RECEIVED;
```
Comment (samuel40791765) [inline ssl/extensions.cc]:
```suggestion
  hs->ssl->s3->pha_ext = SSL_PHA_EXT_RECEIVED;
```

File: ssl/extensions.cc [hunk 11]
```diff
@@ -3089,6 +3089,48 @@ bool ssl_negotiate_alps(SSL_HANDSHAKE *hs, uint8_t *out_alert,
...
+static bool ext_pha_parse_clienthello(SSL_HANDSHAKE *hs,
+                                      uint8_t *out_alert,
+                                      CBS *contents)  {
+  if(contents == NULL || hs->min_version < TLS1_3_VERSION) {
+    return true;
+  }
+
+  if(CBS_len(contents) == 0) {
+    return false;
+  }
```
Comment (samuel40791765) [inline ssl/extensions.cc]: This seems a bit different from what OpenSSL has, is this intentional? 
https://github.com/openssl/openssl/blob/master/ssl/statem/extensions_srvr.c#L1248-L1252

File: ssl/internal.h [hunk 12]
```diff
@@ -3878,6 +3882,10 @@ struct ssl_ctx_st {
   // of support for AES hardware. The value is only considered if
   // |aes_hw_override| is true.
   bool aes_hw_override_value : 1;
+
+  // State of post handshake authentication. Can be any state from SSL_PHA_STATE
+  uint8_t pha_ext = 0;
```
Comment (samuel40791765) [inline ssl/internal.h]: `SSL_CTX` should only be for global configuration settings of the connection. In other words, `SSL_CTX` only controls whether pha is enabled. The actual processing state should be up to `SSL3_STATE` to maintain.

File: ssl/tls13_server.cc [hunk 13]
```diff
@@ -826,7 +826,14 @@ static enum ssl_hs_wait_t do_send_server_hello(SSL_HANDSHAKE *hs) {
...
   if (!ssl->s3->session_reused) {
     // Determine whether to request a client certificate.
-    hs->cert_request = !!(hs->config->verify_mode & SSL_VERIFY_PEER);
+    hs->cert_request = !!(hs->config->verify_mode & SSL_VERIFY_PEER) &&
+                       !(hs->config->verify_mode & SSL_VERIFY_POST_HANDSHAKE);
+
+    // Client authentication requested but immediately after initial handshake
+    if(!!(hs->config->verify_mode & SSL_VERIFY_POST_HANDSHAKE) &&
+        !!(hs->config->verify_mode & SSL_VERIFY_PEER)) {
+      ssl->s3->pha_ext = SSL_PHA_REQUEST_PENDING;
```
Comment (samuel40791765) [inline ssl/tls13_server.cc]: * I'm not quite sure if the original logic in `!!(hs->config->verify_mode & SSL_VERIFY_PEER)` was intentional, or if we could just use `(hs->config->verify_mode & SSL_VERIFY_PEER)` 
* Could you help clarify what the behavior is supposed to be in this if statement? I blame the double exclamation marks, but I'm a bit confused :(  Is this supposed to be configured if both `SSL_VERIFY_POST_HANDSHAKE` and `SSL_VERIFY_PEER` are on?

**Review by samuel40791765 — COMMENTED** [review]
I've only briefly looked over the tests for now, but they seem solid :) I'll take a closer look at them in the next iteration

File: include/openssl/ssl.h (hunk 2 above)
Comment (smittal3) [reply include/openssl/ssl.h]: Added in!

File: include/openssl/ssl.h (hunk 3 above)
Comment (smittal3) [reply include/openssl/ssl.h]: Ah, I had put in bool for default but must not have changed it since I haven't implemented the functionality for sending the CertificateRequest yet. I will remove this functionality from this commit for now but I have made the changes.

File: ssl/extensions.cc (hunk 11 above)
Comment (smittal3) [reply ssl/extensions.cc]: I saw CBB_add_u16_length_prefixed being used to specify the length of the contents field but in our case I guess the contents field is always 0. Updated both the parse and add functions accordingly.

File: include/openssl/ssl.h (hunk 6 above)
Comment (smittal3) [reply include/openssl/ssl.h]: Ahhh gotcha, will keep in mind for future additions!

File: ssl/ssl_lib.cc (hunk 7 above)
Comment (smittal3) [reply ssl/ssl_lib.cc]: Good call, I didn't make that distinction. Have added in another field to manage that.

File: ssl/extensions.cc (hunk 8 above)
Comment (smittal3) [reply ssl/extensions.cc:3106]: If adding an extension returns false, we also throw an error. This just happens in the calling context instead of the function itself: https://github.com/smittal3/aws-lc/blob/4b5c619d7a326ecd673ae6afac2774ad8fc0bb97/ssl/extensions.cc#L3470

File: ssl/tls13_server.cc (hunk 13 above)
Comment (smittal3) [reply ssl/tls13_server.cc]: - The double exclamation converts a given value into the same truth value in a boolean format (ex. taking an int and then applying the !! operator will give you a boolean representing the same truth value). In this context, the & operator returns the binary addition of the two values provided (which could be any int, 10 & 12 would return 8). I assume the !! is just to explicitly ensure we are converting the result into a boolean value.
- The behavior before was if SSL_VERIFY_PEER is set, then hs->cert_request should be true. This makes sense since if the server sets this flag, they want to send a certificate request in the handshake. With the introduction of SSL_VERIFY_POST_HANDSHAKE, we allow servers to modify this behavior. If both flags are set together, then the server will authenticate the client but right after the initial handshake. So here, only when both flags are set, we set pha_ext to SSL_PHA_REQUEST_PENDING and leave hs->cert_request to false such that the server does not send a certificate request in the initial handshake but does so right after it. 
In short: If both are set, then do PHA immediately after initial handshake. If only SSL_VERIFY_PEER is set, we do the default behavior from before.

File: ssl/internal.h [hunk 14]
```diff
@@ -2804,6 +2813,12 @@ struct SSL3_STATE {
   // received.
   uint8_t warning_alert_count = 0;
 
+
+  // State of post handshake authentication. Can be any state from SSL_PHA_STATE
+  uint8_t pha_ext = 0;
+  // Holds whether PHA is enabled, by default it is not (value 0)
+  bool pha_enabled = 0;
```
Comment (samuel40791765) [inline ssl/internal.h]: * Standard way would be to define this as `bool pha_enabled : 1;`, then initialize it as `false` in the constructor.
https://github.com/smittal3/aws-lc/blob/4b5c619d7a326ecd673ae6afac2774ad8fc0bb97/ssl/s3_lib.cc#L166-L181

* Our documentation follows a standard pattern with the variable name first. The other variable documentation in this file would server as good example.
ex: `// pha_enabled indicates whether PHA is enabled. It is off by default.`

File: ssl/internal.h [hunk 15]
```diff
@@ -2804,6 +2813,12 @@ struct SSL3_STATE {
   // received.
   uint8_t warning_alert_count = 0;
 
+
+  // State of post handshake authentication. Can be any state from SSL_PHA_STATE
+  uint8_t pha_ext = 0;
```
Comment (samuel40791765) [inline ssl/internal.h]: * Can we redefine this as `SSL_PHA_STATE ssl_pha_state = SSL_PHA_NONE`? The enum makes things clearer with what available states we have. Similar to what we do below with `ssl_ech_status_t ech_status = ssl_ech_none;`

* Minor spacing nit:
```suggestion
  // pha_ext indicates the state of post handshake authentication. It can
  // be any state from |SSL_PHA_STATE|
  uint8_t pha_ext = 0;
  
```
This follows the rest of the codebase documentation format.

File: ssl/internal.h [hunk 16]
```diff
@@ -3878,6 +3893,10 @@ struct ssl_ctx_st {
   // of support for AES hardware. The value is only considered if
   // |aes_hw_override| is true.
   bool aes_hw_override_value : 1;
+
+  // State of post handshake authentication, by default it's disabled (value 0)
+  bool pha_enabled = 0;
```
Comment (samuel40791765) [inline ssl/internal.h]: Same thing here: https://github.com/smittal3/aws-lc/blob/main/ssl/ssl_lib.cc#L529

File: ssl/ssl_lib.cc [hunk 17]
```diff
@@ -929,6 +931,24 @@ static int ssl_do_post_handshake(SSL *ssl, const SSLMessage &msg) {
   return 1;
 }
 
+// Sets pha_enabled value in SSL_CTX, 0 is disabled, 1 is enabled
+void SSL_CTX_set_post_handshake_auth(SSL_CTX *ctx, int val) {
+  ctx->pha_enabled = val;
+}
+
+// Sets pha_enabled value in SSL, 0 is disabled, 1 is enabled
```
Comment (samuel40791765) [inline ssl/ssl_lib.cc]: Minor Nit: Comments here are a bit redundant, since standard C++/C indicate 1 for true and 0 for false. :)

File: ssl/tls13_server.cc (hunk 13 above)
Comment (samuel40791765) [reply ssl/tls13_server.cc]: * Thanks for the great explanation! That really clears up some things. It might be helpful to add some parts of your explanation  above `hs->cert_request` to indicate that we're doing here, and why we leave `cert_request` false if `SSL_VERIFY_POST_HANDSHAKE` is not configured.

* Since we're not assigning the converted value in the if statement, are the double exclamation marks in there redundant? It would make the logic much more easier to read. Is it possible to just use do something like this:
```
if(hs->config->verify_mode & SSL_VERIFY_PEER && hs->config->verify_mode & SSL_VERIFY_POST_HANDSHAKE)
```
I noticed it's what we do in other places of our code: https://github.com/aws/aws-lc/blob/7bdae649ee6a723a5197e27ec3770c37f6a64e85/ssl/tls13_server.cc#L831

File: ssl/ssl_test.cc [hunk 18]
```diff
@@ -8707,6 +8707,163 @@ TEST(SSLTest, ProcessTLS13NewSessionTicket) {
...
+  EXPECT_EQ(server.get()->s3->pha_ext, SSL_PHA_NONE);
+
+  // Enable pha_ext for client
+  SSL_set_post_handshake_auth(client.get(), 1);
+  EXPECT_EQ(client_ctx.get()->pha_enabled, 0);
+  EXPECT_EQ(server_ctx.get()->pha_enabled, 0);
+  EXPECT_EQ(client.get()->s3->pha_enabled, 1);
+  EXPECT_EQ(server.get()->s3->pha_enabled, 0);
+  EXPECT_EQ(client.get()->s3->pha_ext, SSL_PHA_NONE);
+  EXPECT_EQ(server.get()->s3->pha_ext, SSL_PHA_NONE);
```
Comment (samuel40791765) [inline ssl/ssl_test.cc:8992]: Since you're doing this in each test, just with different flags, it might help to refactor this into a function or a test suite class constructer.

**Review by samuel40791765 — COMMENTED** [review]
Test cases look solid :) 
You might want to rebase the branch you're pushing to, since the file diff is getting muddled with our commits on the main repo.

File: include/openssl/ssl.h [hunk 19]
```diff
@@ -4501,6 +4506,32 @@ OPENSSL_EXPORT int SSL_CTX_set_max_send_fragment(SSL_CTX *ctx,
 OPENSSL_EXPORT int SSL_set_max_send_fragment(SSL *ssl,
                                              size_t max_send_fragment);
 
+
+
+// SSL_CTX_set_post_handshake_auth and SSL_set_post_handshake_auth enable the
```
Comment (justsmth) [inline include/openssl/ssl.h:4511]: Nit: Your changes stand out as your lines go past column 80.  Try to keep all lines to 80 characters or shorter.

You can try using clang-format to help with this:
```
clang-format -i include/openssl/ssl.h
```

Be careful, clang-format can quickly make sweeping changes and add a lot of unnecessary "noise" to code diffs.

File: include/openssl/ssl.h [hunk 20]
```diff
@@ -4501,6 +4506,32 @@ OPENSSL_EXPORT int SSL_CTX_set_max_send_fragment(SSL_CTX *ctx,
...
+// The |SSL_VERIFY_PEER| flag must be set; the |SSL_VERIFY_POST_HANDSHAKE| flag is optional.
+// These flags can be set via |SSL_CTX_set_verify| for |SSL_CTX| or |SSL_set_verify| for |SSL|.
+OPENSSL_EXPORT int SSL_verify_client_post_handshake(SSL *ssl);
+
+
+// SSL_get_verify_result_pha returns the status of post handshake authentication. It
+// returns |X509_V_OK| if the peer certificate is provided and valid, and NULL otherwise.
+// The function calls |SSL_get_peer_certificate| to ensure a peer certificate has been provided
+// and |SSL_get_verify_result| to ensure the provided certificate is valid.
+    OPENSSL_EXPORT long SSL_get_verify_result_pha(const SSL *ssl);
```
Comment (justsmth) [inline include/openssl/ssl.h]: Nit: This line should not be indented.

File: ssl/extensions.cc [hunk 21]
```diff
@@ -3089,6 +3089,46 @@ bool ssl_negotiate_alps(SSL_HANDSHAKE *hs, uint8_t *out_alert,
...
 }
 
+// Adds pha_ext to the client hello
+static bool ext_pha_add_clienthello(const SSL_HANDSHAKE *hs, CBB *out,
+                                    CBB *out_compressible, ssl_client_hello_type_t type)  {
+  const SSL *const ssl = hs->ssl;
+
+  // post_handshake_auth is not supported in TLS versions < 1.3. Also checking
+  // if extension has been enabled by client.
+  if(hs->min_version < TLS1_3_VERSION || ssl->s3->pha_enabled != 1) {
```
Comment (justsmth) [inline ssl/extensions.cc]: Should the extension be restricted based on the `min_version` or `max_version`?

ALPS also requires TLS 1.3, but its logic considered the `max_version`:
```
  if (// ALPS requires TLS 1.3.
      hs->max_version < TLS1_3_VERSION ||
```

File: ssl/extensions.cc [hunk 22]
```diff
@@ -3089,6 +3089,46 @@ bool ssl_negotiate_alps(SSL_HANDSHAKE *hs, uint8_t *out_alert,
...
+                                      uint8_t *out_alert,
+                                      CBS *contents)  {
+  if(contents == nullptr || hs->min_version < TLS1_3_VERSION) {
+    return true;
+  }
+
+  // Should be empty
+  if(CBS_len(contents) != 0) {
+    return false;
+  }
```
Comment (justsmth) [inline ssl/extensions.cc]: What are the semantics of `true` and `false` here?

File: ssl/internal.h (hunk 15 above)
Comment (justsmth) [reply ssl/internal.h]: I agree -- use the enum you defined for this.

File: ssl/internal.h (hunk 14 above)
Comment (justsmth) [reply ssl/internal.h]: Is the value of `pha_enabled` implied by the value of `pha_ext`. Could `SSL_PHA_DISABLED` be an enum value?

File: ssl/ssl_lib.cc [hunk 23]
```diff
@@ -696,6 +696,8 @@ SSL *SSL_new(SSL_CTX *ctx) {
     return nullptr;
   }
 
+  ssl->s3->pha_enabled = ctx->pha_enabled;
```
Comment (justsmth) [inline ssl/ssl_lib.cc:700]: Does this need to also update/initialize the value of `ssl->s3->pha_ext`?

File: ssl/ssl_lib.cc [hunk 24]
```diff
@@ -929,6 +931,24 @@ static int ssl_do_post_handshake(SSL *ssl, const SSLMessage &msg) {
...
 }
 
+// Sets pha_enabled value in SSL_CTX, 0 is disabled, 1 is enabled
+void SSL_CTX_set_post_handshake_auth(SSL_CTX *ctx, int val) {
+  ctx->pha_enabled = val;
+}
+
+// Sets pha_enabled value in SSL, 0 is disabled, 1 is enabled
+void SSL_set_post_handshake_auth(SSL *ssl, int val) {
+  ssl->s3->pha_enabled = val;
```
Comment (justsmth) [inline ssl/ssl_lib.cc:940]: Does this need to also update/initialize the value of `ssl->s3->pha_ext`?

File: ssl/tls13_server.cc (hunk 13 above)
Comment (smittal3) [reply ssl/tls13_server.cc]: I'll add in some of this explanation explicitly in the comments. As for the exclamation marks I agree, they make the logic confusing. I wasn't sure if they were put in for a reason, but I can simplify it since all we care about is the value not whether it's truly boolean or not.

File: ssl/internal.h (hunk 15 above)
Comment (smittal3) [reply ssl/internal.h]: That does make things clearer, made the change

File: ssl/internal.h (hunk 16 above)
Comment (smittal3) [reply ssl/internal.h]: Done!

File: ssl/extensions.cc (hunk 21 above)
Comment (smittal3) [reply ssl/extensions.cc]: min_version indicates the minimum accepted and max_version indicates maximum accepted version. It could be the case that min_version is not 1.3 so you're correct, we should check against max_version. But I'm not sure if that's enough to ensure that 1.3 is being negotiated. Something to keep in mind for testing.

File: ssl/ssl_lib.cc (hunk 23 above)
Comment (smittal3) [reply ssl/ssl_lib.cc:700]: I'm not directly modifying that value here. pha_enabled holds whether or not the functionality is being turned on. Based on it's value, when we actually add the extension to the client hello (if enabled), that's when pha_ext gets assigned SSL_PHA_EXT_SENT. And then SSL_PHA_EXT_RECEIVED when the server parses it from the client hello and so on.

**Review by smittal3 — COMMENTED** [review]
responded to comments

File: ssl/internal.h (hunk 14 above)
Comment (smittal3) [reply ssl/internal.h]: pha_enabled is what is set initially. pha_ext holds the state for what part of PHA we are in (extension sent, received, pending, etc.). pha_enabled is used as a check to ensure we should be processing and changing value of pha_ext. pha_ext has a default state called SSL_PHA_NONE which indicates we are not conducting PHA. 
So the value of pha_enabled is not implied by pha_ext. Whether we conduct PHA and transition states in pha_ext is initially checked via pha_enabled.

File: ssl/extensions.cc (hunk 22 above)
Comment (smittal3) [reply ssl/extensions.cc]: If false, the calling context returns an error. False just indicates parsing had an error. So in the above code snippet, if the body of the extension is not of length 0 (which it should be since we constructed it that way in ext_pha_add_clienthello), we'd return false and the calling context (https://github.com/smittal3/aws-lc/blob/4b5c619d7a326ecd673ae6afac2774ad8fc0bb97/ssl/extensions.cc#L3661) would return an error. Otherwise true indicates successful parsing.
